### PR TITLE
feature(application): adds services to Application and adds elgg()

### PR DESCRIPTION
**Pull request by mrclay**
elgg() provides access to the booted Application instance which will provide access to future public API service objects (instead of global functions).

As a first step, this adds only the “config” service, which is needed to use the method getVolatile() before site boot.

For #8238

### CODE REVIEW

File: engine/classes/Elgg/Application.php [hunk 1]
```diff
@@ -27,6 +32,18 @@ class Application {
...
+	 *
+	 * @var string[]
+	 */
+	private static $public_services = [
+		'config',
+		'context',
+		'events',
+		'hooks',
+	];
+
```
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: These were just a few starters. So many decisions to make...

File: engine/classes/Elgg/Application.php [hunk 2]
```diff
@@ -183,6 +175,8 @@ public function loadCore() {
 				}
 			}
 
+			elgg($this);
```
Comment (ewinslow) [inline engine/classes/Elgg/Application.php]: this line is rather unclear... what in the world does this do?

File: engine/tests/phpunit/Elgg/ApplicationTest.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		$services = _elgg_services();
+		$app = new Application($services);
+
+		$this->assertSame($services->config, $app->config);
+		$this->assertSame($services->context, $app->context);
+		$this->assertSame($services->events, $app->events);
+		$this->assertSame($services->hooks, $app->hooks);
+	}
+
+	function testLoadSettings() {
```
Comment (ewinslow) [inline engine/tests/phpunit/Elgg/ApplicationTest.php]: test names should also include expected result.

**Comment by ewinslow** [comment]
I like the direction. Wondering if we should skip to php-di instead and map from interface names to instances rather than string keys to instances?

**Comment by ewinslow** [comment]
I.e.:

```php

use Elgg\Context;

elgg()->get(Context::class)->push('foo');
```

as opposed to:

```php
elgg()->context->push('foo');
```

The latter is certainly more terse but it doesn't have the nice, automatic namespacing benefit that comes with using classes/interfaces as keys.

**Comment by ewinslow** [comment]
Obviously in older version of PHP that would have to be `elgg()->get('Elgg\Context')->push('foo');`

File: engine/classes/Elgg/Application.php (hunk 2 above)
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: Why it Elgg's this thing obviously! Will add comment. :)

**Comment by mrclay** [comment]
What's the "namespacing benefit"? The huge advantage of faux properties/methods is IDE static analysis/autocomplete. Really we could support both syntaxes easily like we'd planned to before.

**Comment by mrclay** [comment]
Having explicit core class names in 3rd party plugin code seems like a downside to me, considering I can jump straight to the method definition in the IDE. If plugins don't bind to class names, that seems more flexible for core.

File: engine/classes/Elgg/Application.php (hunk 2 above)
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application.php]: :)

**Comment by mrclay** [comment]
Latest push adds "session" and "site" (both already accessible from current API) and uses an Application static prop to simplify elgg().

**Comment by ewinslow** [comment]
> Having explicit core class names in 3rd party plugin code seems like a downside to me

This is not something we'll be able to avoid when we start supporting ctor/method injection. But that still comes with ide support also, so I'm fine with this for now.

**Comment by ewinslow** [comment]
If we do start leaking core class names into plugin code, we should strive to make this only be interface names, rather than concrete classes...

**Comment by Srokap** [comment]
> If we do start leaking core class names into plugin code, we should strive to make this only be interface names, rather than concrete classes...

Very much agreed.

**Comment by mrclay** [comment]
While the Application class is a step forward, I'm less excited about the process of slowly opening up existing services, making interfaces, etc. As I've blogged about before, this is a huge undertaking that is a big distraction away from more high value work we could be doing on the front end.

I'd suggest we reserve elgg() for newly created APIs rather than making duplicates of existing APIs.

**Comment by ewinslow** [comment]
Speaking of front end, I have some new ideas there I'm really excited about... Will post elsewhere...

**Comment by juho-jaakkola** [comment]
> I'd suggest we reserve elgg() for newly created APIs rather than making duplicates of existing APIs.

Not counting the ones already added to this PR?

**Comment by mrclay** [comment]
I think it would be messy to make an arbitrary chunk of our API accessible two ways. It widens the surface area we have to support and document, and these services weren't really designed for dev usability, just quickly copied to classes.

So yes I'd be in favor of removing the pre-existing services I've exposed here.

**Comment by mrclay** [comment]
I've rolled this back a bit to just introduce elgg() and provide the "config" service, which was already exposed anyway in the master branch. It doesn't add docs as there's no reason yet for devs to use it.

**Comment by ewinslow** [comment]
Sgtm

**Comment by mrclay** [comment]
FYI, in #8288 I've changed the "config" service to be exposed as an interface, and documented services and service creation.

**Pull request merged**